Improve code coverage CI to support Fedora python3

cathay4t opened the pull request:
Changes the hard coded python site path to use rpm macro output by check
whether we are running on Fedora or Centos.
Let the Fedora docker instance run both integration tests and tox tests
in order to gather all code coverage reports.

EdDev commented:
There are multiple points handled in this change.

Generalizing the python site-packages for py2 and py3 is very nice.

The other/s, are trying to merge the unit and integration tests, which requires good convincing. I personally do not understand why is that good. It is not supposed to be in the same CI job as far as I am concerned (this is the case now).

tyll commented:
The main idea is to run the unit tests in a Fedora container instead of running them on Travis' Ubuntu installation. Then we do not have to handle the unit tests special anymore (see #214). Having the tests in the same CI job is a (temporary?) result from this but not the goal

EdDev commented on automation/run-tests.sh:
```diff
@@ -27,7 +32,11 @@ function pyclean {
 function docker_exec {
-    docker exec $USE_TTY -i $CONTAINER_ID /bin/bash -c "$1"
+    docker exec \
+        -e TRAVIS="$TRAVIS" \
```
Can these be an input to this function?

EdDev commented on automation/run-tests.sh:
```diff
@@ -27,7 +32,11 @@ function pyclean {
+    docker exec \
+        -e TRAVIS="$TRAVIS" \
+        -e TRAVIS_JOB_ID="$TRAVIS_JOB_ID" \
+        -e TRAVIS_BRANCH="TRAVIS_BRANCH" \
```
Did you mean `$TRAVIS_BRANCH` ?

EdDev commented:
I also suggest focus the commits/PR on the change, not the end-goal. You can fix the paths of py2/3, suggest improvements for integrating with travis-ci and try to suggest the tests merging. Each is an independent change, together they eventually increase coverage. (the long goal may be mentioned in the commit message if you like)

tyll commented:
> Locally, I want to run `tox` and the integration tests independently.

You can still do this AFAICS.

> I am still missing the good reasoning to do this merging of tests.
> Why is the unit tests dependent on the container image? These tests are supposed to be OS agnostic and tox is taking care of the dependencies in the env. Specifically, they are testing the sources and not the installed rpm packages, or in other words, it should not require deployment in order to run.

The test setup is not OS agnostic, as you can see in https://travis-ci.org/nmstate/nmstate/builds/480421674#L471 tox installs nmstate for the unit tests. Currently this only works because nmstate does not depend (via metadata) on PyGObject even though the code does. This can be reproduced with using a fresh CentOS cloud image, for example CentOS-7-x86_64-GenericCloud-1804_02.qcow2c

Preparation:
```
yum remove -y python-gobject-base  # this is a dependency of tuned, so it is already present on the image
yum install -y epel-release
yum install -y python2-pip
pip install nmstate
```

Test:
```
# nmstatectl show
Traceback (most recent call last):
  File "/usr/bin/nmstatectl", line 9, in <module>
    load_entry_point('nmstate==0.0.3', 'console_scripts', 'nmstatectl')()
  File "/usr/lib/python2.7/site-packages/nmstatectl/nmstatectl.py", line 54, in main
    return args.func(args)
  File "/usr/lib/python2.7/site-packages/nmstatectl/nmstatectl.py", line 125, in show
    state = _filter_state(netinfo.show(), args.only)
  File "/usr/lib/python2.7/site-packages/libnmstate/netinfo.py", line 34, in show
    report = {Constants.INTERFACES: interfaces()}
  File "/usr/lib/python2.7/site-packages/libnmstate/netinfo.py", line 55, in interfaces
    for dev in nm.device.list_devices()]
  File "/usr/lib/python2.7/site-packages/libnmstate/nm/device.py", line 325, in list_devices
    client = nmclient.client(refresh=True)
  File "/usr/lib/python2.7/site-packages/libnmstate/nm/nmclient.py", line 48, in client
    _nmclient = NM.Client.new(None)
AttributeError: 'NoneType' object has no attribute 'Client'
```

This is actually an ImportError which is masked for unit tests in nmclient.py:
```
try:
    import gi                           # pylint: disable=import-error
    gi.require_version('NM', '1.0')     # NOQA: F402
    from gi.repository import Gio, GLib, NM  # pylint: disable=import-error
    from gi.repository import GObject        # pylint: disable=import-error
except ImportError:
    # Allow the error to pass in case we are running in a unit test context
    g = None
    Gio = None
    GLib = None
    GObject = None
    NM = None
```

Therefore I would argue that it would even be better to properly install the dependencies for the unit tests to avoid these kind of hacks which lead to strange error messages when not  running a unit test.

cathay4t commented:
> Yes, in the way it is defined today it does. It worked well as long as things are mocked correctly. If now we run non unit tests or do not mock dependencies, it is something that needs fixing.

We have user-facing problem: `pip install nmstate cannot be run as missing dependency on gobject`.
And you are talking about dev-facing design: `tox is OS independent and should not be run in docker for test`.

Solving the user problem take priority, if you have good way to solve the missing gobject dependency with CI all pass, please enlighten us.

EdDev commented:
@cathay4t , pip depends on python packages (only), your rpm build is dependent on setup.py which is dependent on the requirements.txt, which now does not contain all dependencies.

If the user cannot install nmstate from pip only, I do not see how you solve it here.

I'm also not familiar with any project that requires full installation before running its unit tests, it contradicts the development flow.

tyll commented:
> > The test setup is not OS agnostic,
> 
> My claim is that unit tests should be agnostic to the OS. A developer does not need to install nmstate in order to develop it.

The developer does not need to install nmstate but only a container runtime environment with this  full change. If they would not like to run the container than installing a recent version of gobject-introspection-devel is reasonable to me.

> > tox installs nmstate for the unit tests.
> 
> Yes, in the way it is defined today it does. It worked well as long as things are mocked correctly. If now we run non unit tests or do not mock dependencies, it is something that needs fixing.

I do not see that the effort to have developers not install gobject-introspection-devel in a reasonable version is worth figuring out workarounds for this.

EdDev commented:
> The developer does not need to install nmstate but only a container runtime environment with this full change. If they would not like to run the container than installing a recent version of gobject-introspection-devel is reasonable to me.

Why not mention it in the tox.ini file? Won't that solve it?

tyll commented:
> @cathay4t , pip depends on python packages (only), your rpm build is dependent on setup.py which is dependent on the requirements.txt, which now does not contain all dependencies.

Not having all requirements for running nmstate in requirements.txt is a packaging bug.

> If the user cannot install nmstate from pip only, I do not see how you solve it here.

I tried to fix this in #214 but the CI does not support thi.

> I'm also not familiar with any project that requires full installation before running its unit tests, it contradicts the development flow.

The developer does not need to install nmstate, only tox installs it (by default) in their virtenv. IMHO this is a strong indication that other projects also install their package in tox, otherwise it would not be the default.

> > The developer does not need to install nmstate but only a container runtime environment with this full change. If they would not like to run the container than installing a recent version of gobject-introspection-devel is reasonable to me.
> 
> Why not mention it in the tox.ini file? Won't that solve it?

It is not a python package, therefore it cannot be mentioned in the tox.ini file - if it can, it would be a good idea as well.

cathay4t commented:
V2 patch set uploaded.

Changes since V1:
 * Split to two patches: one for code coverage for python3, one for upload coverage data in container. Let me know if you want me to do 1 patch per PR.

tyll commented on .coveralls.yml:
```diff
@@ -0,0 +1,3 @@
+repo_token: Fc9rievh0XnhuLbMs3cMqi7rDcvQhLzLw
```
This needs to be removed and re-generated. Then it should be added to travis as a environment variable `COVERALLS_REPO_TOKEN`. It is not meant to be public.

tyll commented on automation/run-integration-tests.sh:
```diff
@@ -31,7 +31,11 @@ function pyclean {
+    docker exec \
+        -e TRAVIS="$TRAVIS" \
+        -e TRAVIS_JOB_ID="$TRAVIS_JOB_ID" \
+        -e TRAVIS_BRANCH="$TRAVIS_BRANCH" \
```
This should also pass `COVERALLS_REPO_TOKEN` to the container.

cathay4t replied:
> Please protect the COVERALLS_REPO_TOKEN.

What's the difference by putting that key to `.travis.yml` vs `..coveralls.yml`? Both no protection.

Or maybe you mean other way.

tyll replied:
It is possible to set environment variables at https://travis-ci.org/nmstate/nmstate/settings

tyll replied:
Or better see my other comment about running coveralls outside of the container.

cathay4t replied:
Good, I will use it.

tyll commented on automation/run-integration-tests.sh:
```diff
@@ -66,11 +74,14 @@ function run_tests {
         tests/integration \
     ${nmstate_pytest_extra_args}"
+    if [ -n "$TRAVIS" ];then
+        docker_exec 'cd /workspace/nmstate && coveralls'
```
Does coveralls really need to run in the container? It might also be possible to run it from outside the container since the coverage report should be available there as well since it is a shared volume.

cathay4t replied:
The reason to run it in docker is `coveralls` need access to the source of `libnmstate` where
outside docker does not. Sorry I forgot to mention that in commit message.

cathay4t commented:
@tyll Even I stored COVERALLS_REPO_TOKEN in traivs-ci config, I still need to pass it to container via `docker exec -e COVERALLS_REPO_TOKEN="$COVERALLS_REPO_TOKEN"` which will expose it in log.

I don't think COVERALLS_REPO_TOKEN is that sensitive, can I still use the `.coveralls.yml` way?

cathay4t commented:
Rebased and amend the commit message to include reason why we need to run `coveralls` in docker.

tyll commented:
> @tyll Even I stored COVERALLS_REPO_TOKEN in traivs-ci config, I still need to pass it to container via `docker exec -e COVERALLS_REPO_TOKEN="$COVERALLS_REPO_TOKEN"` which will expose it in log.
> 
> I don't think COVERALLS_REPO_TOKEN is that sensitive, can I still use the `.coveralls.yml` way?

AFAIU, everyone with access to the  COVERALLS_REPO_TOKEN can upload wrong coverage data for our project. Not sure, if there is someone who would really benefit from this but it is not correct.

It seems a workaround to upload the data from outside the container is:

```
mv .coverage .coverage.orig
cat > .coveragerc <<EOF
[paths]
source =
    .
    /usr/lib/python2.7/site-packages
    /usr/lib/python3.6/site-packages
    /usr/lib/python3.7/site-packages
EOF
coverage combine --append --rcfile .coveragerc
```

Afterwards the sources are correct in the .coverage file. AFAIU with the current setup, the coverage report contains only the last coverage data, otherwise we need to pass `--cov-append` to tox. Not sure, what kind of coverage data do we want? IMHO everything combined - @EdDev ?

cathay4t commented:
V3 patch uploaded.

Changes since V2:
 * Removed the patch for uploading coverage data in container patch.
 * No changes was made to V2 patch on python3 code coverage.
 * The code coverage patch will be included in other PR.

I managed to both split and combine the code coverage:

https://coveralls.io/builds/21304281

In above build, you may found out code coverage for unit test and integration test separated.

cathay4t commented:
> > Previously, we are hard coded python site path to python2 one,
> > this patch change it to `rpm -E %{python_sitelib}` for Centos docker
> > image and `rpm -E %{python3_sitelib}` for Fedora docker image.
> 
> Python site packages path is currently fixed to py2. In order to support py3, set the path based on the image.

To hard code python3 path also requires us to update that path every time Fedora update python3 version
Using rpm macro is more fitting as we are using rpm package for testing, if we hard coded the path,
we might hit a problem if Fedora/Redhat decided to change that path.

cathay4t commented:
> Thanks for the clarification, I fully agree with the fix.
> 
> My point was to request a rephrase of the commit message. You may or may not add details on how you implemented the fix.

Then what should I input in commit message?

tyll commented:
> I managed to both split and combine the code coverage:
> 
> https://coveralls.io/builds/21304281

I can only find the split coverage but not the combined coverage, where is the combined coverage?

cathay4t commented:
> > I managed to both split and combine the code coverage:
> > https://coveralls.io/builds/21304281
> 
> I can only find the split coverage but not the combined coverage, where is the combined coverage?

At the `Run Details`:

3500 of 5316 relevant lines covered (65.84%) 0.66 hits per line

tyll commented:
> At the `Run Details`:
> 
> 3500 of 5316 relevant lines covered (65.84%) 0.66 hits per line

There merged coverage needs to be at least as high as the highest single coverage (the integration tests  have 86% coverage), therefore it appears to me that this is just the coverage from the last job.

EdDev commented:
>     My point was to request a rephrase of the commit message. You may or may not add details on how you implemented the fix.
> 
> Then what should I input in commit message?

That was in my first comment: "Python site path is now set based on the container image: ... for CentOS, ... for Fedora."

tyll commented:
> That was in my first comment: "Python site path is now set based on the container image: ... for CentOS, ... for Fedora."

Actually it is based on the python version with using python2 on CentOS and python3 on Fedora and the minor version is determined using RPM.

EdDev commented:
> 
> 
>     That was in my first comment: "Python site path is now set based on the container image: ... for CentOS, ... for Fedora."
> 
> Actually it is based on the python version with using python2 on CentOS and python3 on Fedora and the minor version is determined using RPM.

I personally do not need that level of detail, it is written in the code. But I do need a clear explanation on what was done in a simple form. @cathay4t can you please update the commit message with or without mentioned details?

cathay4t commented:
@EdDev Commit message updated.

tyll commented:
ci test please

tyll commented:
ci test please
